setup.go

```diff
@@ -0,0 +1,828 @@
+			state, err = opts.askServerURL(ctx, stdin)
+
+		case stateServerIP:
+			state, err = opts.askServerIP(ctx, stdin)
+
+		case stateServerCert:
+			state, err = opts.askServerCert(ctx, stdin)
+
+		case stateHostedMenderCredentials:
+			state, err = opts.askHostedMenderCredentials(ctx, stdin)
+
+		case statePolling:
+			state, err = opts.askPollingIntervals(ctx, stdin)
+		}
+		if err != nil {
+			return err
+		}
+	} // END for {state}
+	return opts.saveConfigOptions(config)
+}
+
+func (opts *setupOptionsType) saveConfigOptions(config *menderSysConfig) error {
+	if opts.demo {
+		config.InventoryPollIntervalSeconds = demoInventoryPoll
+		config.RetryPollIntervalSeconds = demoRetryPoll
+		config.UpdatePollIntervalSeconds = demoUpdatePoll
+	} else {
+		config.InventoryPollIntervalSeconds = opts.invPollInterval
+		config.UpdatePollIntervalSeconds = opts.updatePollInterval
+		config.RetryPollIntervalSeconds = opts.retryPollInterval
+	}
+	config.HttpsClient.Certificate = opts.serverCert
+	config.TenantToken = opts.tenantToken
+
+	// Make sure serverURL and devicetypefile is set
+	config.Servers = []client.MenderServer{
+		{
+			ServerURL: opts.serverURL},
+	}
+	if config.DeviceTypeFile == "" {
+		config.DeviceTypeFile = "/etc/mender/device_type"
+	}
+	if config.ClientProtocol == "" {
+		config.ClientProtocol = "https"
+	}
+
+	if err := storeConfigFile(config, opts.configPath); err != nil {
+		return err
+	}
+	err := ioutil.WriteFile(config.DeviceTypeFile,
+		[]byte("device_type="+opts.deviceType), 0644)
+	if err != nil {
+		return errors.Wrap(err, "Error writing to devicefile.")
+	}
+	if opts.demo && !opts.hostedMender {
+		opts.maybeAddHostLookup()
+	}
+	return nil
+}
+
+func (opts *setupOptionsType) maybeAddHostLookup() {
+	f, err := os.OpenFile("/etc/hosts", os.O_RDWR, 0644)
+	if err != nil {
+		log.Warnf("Unable to open \"/etc/hosts\" for appending "+
+			"local route: %s", err.Error())
+		return
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	// Regex: $1: schema, $2: URL, $3: path
+	re, err := regexp.Compile(`(https?://)?(.*)(/.*)?`)
+	if err != nil {
+		log.Warn("Unable to compile regular expression for parsing " +
+			"server URL.")
+		return
+	}
+	// strip schema and path
+	host := re.ReplaceAllString(opts.serverURL, "$2")
+	fmt.Println(host)
+
+	// Check if route allready exists
+	for scanner.Scan() {
+		if strings.Contains(scanner.Text(), host) {
+			return
+		}
+	}
+	route := fmt.Sprintf("\n%-15s %s", opts.serverIP, host)
+
+	// Seek to last character
+	_, err = f.Seek(-1, os.SEEK_END)
+	if err != nil {
+		log.Warnf("Unable to add route %q to \"/etc/hosts\": %s",
+			route, err.Error())
+	}
+	// Remove newline from route string if there allready is one
+	lastChar := make([]byte, 1)
+	if _, err := f.Read(lastChar); err == nil &&
+		lastChar[0] == byte('\n') {
+		route = route[1:]
+	}
+
+	_, err = f.WriteString(route)
+	if err != nil {
+		log.Warnf("Unable to add route %q to \"/etc/hosts\": %s",
+			route, err.Error())
+	}
```

Comment:
I would write an info at the end of this function. Being something that you run as root, it is nice to get notified of system wide changes. If I am evaluating Mender for demo is okey for the setup to modify `/etc/hosts`, but I would for sure want to clean it up later...

Reply:
As I mentioned above, I would rather merge the `stateServerURL` and `stateServerIP` and prompt for an IP or a general URL (with with the schema set) and add this to the `mender.conf` file instead of having to modify `/etc/hosts'.